gcc 13 -Wdangling-pointer build errors

I have a machine which now has gcc 13 and DR fails to build:

```
In file included from /path/dr/git/src/core/unix/../globals.h:310,
                 from /path/dr/git/src/core/unix/signal_private.h:52,
                 from /path/dr/git/src/core/unix/signal.c:45:
/path/dr/git/src/core/unix/signal.c: In function ‘convert_kernel_sigaction_to_old’:
/path/dr/git/src/core/unix/../utils.h:1462:45: error: storing the address of local variable ‘try__state’ in ‘*try__except.try_except_state’ [-Werror=dangling-pointer=]
 1462 |             (try_pointer)->try_except_state = &try__state;                             \
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~
/path/dr/git/src/core/unix/../utils.h:1428:9: note: in expansion of macro ‘TRY’
 1428 |         TRY(try__except, try_statement, EXCEPT(try__except, except_statement)); \
      |         ^~~
/path/dr/git/src/core/unix/signal.c:2209:5: note: in expansion of macro ‘TRY_EXCEPT’
 2209 |     TRY_EXCEPT(dcontext,
      |     ^~~~~~~~~~
/path/dr/git/src/core/unix/../utils.h:1441:30: note: ‘try__state’ declared here
 1441 |         try_except_context_t try__state;                                               \
      |                              ^~~~~~~~~~
/path/dr/git/src/core/unix/../utils.h:1428:9: note: in expansion of macro ‘TRY’
 1428 |         TRY(try__except, try_statement, EXCEPT(try__except, except_statement)); \
      |         ^~~
/path/dr/git/src/core/unix/signal.c:2209:5: note: in expansion of macro ‘TRY_EXCEPT’
 2209 |     TRY_EXCEPT(dcontext,
      |     ^~~~~~~~~~
/path/dr/git/src/core/unix/signal.c:2205:45: note: ‘dcontext’ declared here
 2205 | convert_kernel_sigaction_to_old(dcontext_t *dcontext, old_sigaction_t *os,
      |                                 ~~~~~~~~~~~~^~~~~~~~
```

I believe that is from `-Wdangling-pointer`.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -886,6 +886,12 @@ if (UNIX)
       set(WARN "${WARN} -Wno-stringop-overflow")
     endif (stringop_overflow_avail)
     CHECK_C_COMPILER_FLAG("-Wtype-limits" HAVE_TYPELIMITS_CONTROL)
+    CHECK_C_COMPILER_FLAG("-Wdangling-pointer" dangling_pointer_avail)
+    if (dangling_pointer_avail)
+      # XXX i#6337: It is difficult to rewrite the TRY macros to avoid this gcc
+      # warning, which is innocuous, so we disable.
+      set(WARN "${WARN} -Wno-dangling-pointer")
+    endif ()
   else (NOT CMAKE_COMPILER_IS_CLANG)
     # clang turns off color when it's writing to a pipe, but the user may still
     # wish to force color if it eventually goes to a terminal.

